### PR TITLE
Optimize CRC intrinisics for targets lacking of CRC extension

### PR DESCRIPTION
**Pull request by Cuda-Chen**
Close https://github.com/DLTcollab/sse2neon/issues/624.

### CODE REVIEW

File: sse2neon.h [hunk 1]
```diff
@@ -8500,12 +8500,8 @@ FORCE_INLINE uint32_t _mm_crc32_u8(uint32_t crc, uint8_t v)
...
 #else
     crc ^= v;
-    for (int bit = 0; bit < 8; bit++) {
-        if (crc & 1)
-            crc = (crc >> 1) ^ UINT32_C(0x82f63b78);
-        else
-            crc = (crc >> 1);
-    }
+    for (int bit = 0; bit < 8; bit++)
+        crc = (crc & 1) ? ((crc >> 1) ^ UINT32_C(0x82f63b78)) : (crc >> 1);
```
Comment (jserv) [inline sse2neon.h:8504]: Don't do this. The optimizing compilers can generate the identical machine code, and the original code is easier to maintain.

Comment (Cuda-Chen) [reply sse2neon.h]: Revert completed.

**Comment by jserv** [comment]
Consider the following changes:
```diff
--- a/sse2neon.h
+++ b/sse2neon.h
@@ -8501,10 +8501,9 @@ FORCE_INLINE uint32_t _mm_crc32_u8(uint32_t crc, uint8_t v)
 #else
     crc ^= v;
     for (int bit = 0; bit < 8; bit++) {
+        crc >>= 1;
         if (crc & 1)
-            crc = (crc >> 1) ^ UINT32_C(0x82f63b78);
-        else
-            crc = (crc >> 1);
+            crc ^= UINT32_C(0x82f63b78);
     }
 #endif
     return crc;
```

File: sse2neon.h [hunk 2]
```diff
@@ -8486,6 +8486,52 @@ FORCE_INLINE uint64_t _mm_crc32_u64(uint64_t crc, uint64_t v)
     return crc;
 }
 
+const uint32_t crc32_tbl[] = {
```
Comment (jserv) [inline sse2neon.h:8489]: Add `static` qualifier and enforce `sse2neon_` prefix.

**Comment by jserv** [comment]
The current `_sse2neon_crc32_tbl` uses 256 * 4 bytes, totaling 1 KiB. Is there a way to reduce its memory usage?

**Comment by Cuda-Chen** [comment]
> Consider the following changes:
> 
> ```diff
> --- a/sse2neon.h
> +++ b/sse2neon.h
> @@ -8501,10 +8501,9 @@ FORCE_INLINE uint32_t _mm_crc32_u8(uint32_t crc, uint8_t v)
>  #else
>      crc ^= v;
>      for (int bit = 0; bit < 8; bit++) {
> +        crc >>= 1;
>          if (crc & 1)
> -            crc = (crc >> 1) ^ UINT32_C(0x82f63b78);
> -        else
> -            crc = (crc >> 1);
> +            crc ^= UINT32_C(0x82f63b78);
>      }
>  #endif
>      return crc;
> ```

This operation does binary division (we would like to check if the LSB is 1, then right-shift 1-bit then XOR, else we just right-shift 1-bit ).
As a result, first right-shifting 1-bit then checking whether the LSB is 1 produces the wrong result.

**Comment by jserv** [comment]
Reference implementation in Linux kernel:
* [arch/arm/crypto/crc32-ce-glue.c](https://github.com/torvalds/linux/blob/master/arch/arm/crypto/crc32-ce-glue.c)
* [arch/arm/crypto/crc32-ce-core.S](https://github.com/torvalds/linux/blob/master/arch/arm/crypto/crc32-ce-core.S)

Accelerated CRC32(C) using ARM CRC, NEON and Crypto Extensions instructions.

File: sse2neon.h [hunk 3]
```diff
@@ -8546,7 +8535,11 @@ FORCE_INLINE uint32_t _mm_crc32_u8(uint32_t crc, uint8_t v)
     crc = __crc32cb(crc, v);
 #else
     crc ^= v;
-    crc = (crc >> 8) ^ _sse2neon_crc32_tbl[crc & 0xFF];
+
+    // Apply half-byte comparision algorithm for the best ratio between
```
Comment (jserv) [inline sse2neon.h:8511]: Since you already shrink the size of the table, move the table inside the function body.

Comment (Cuda-Chen) [reply sse2neon.h:8505]: Done.

File: sse2neon.h [hunk 4]
```diff
@@ -8500,12 +8500,21 @@ FORCE_INLINE uint32_t _mm_crc32_u8(uint32_t crc, uint8_t v)
...
-    for (int bit = 0; bit < 8; bit++) {
-        if (crc & 1)
-            crc = (crc >> 1) ^ UINT32_C(0x82f63b78);
-        else
-            crc = (crc >> 1);
-    }
+
+    // Adapted from: https://create.stephan-brumme.com/crc32/
+    // Apply half-byte comparision algorithm for the best ratio between
+    // performance and look-up table spaces.
```
Comment (jserv) [inline sse2neon.h:8506]: Simply refer it as "lookup table."

**Comment by jserv** [comment]
Reference paper titled [Fast CRC Computation for Generic Polynomials Using PCLMULQDQ Instruction](https://www.lambda-v.com/links/programming/files/intel_white_papers/fast-crc-computation-generic-polynomials-pclmulqdq-paper.pdf).

File: sse2neon.h [hunk 5]
```diff
@@ -8500,12 +8500,47 @@ FORCE_INLINE uint32_t _mm_crc32_u8(uint32_t crc, uint8_t v)
...
+#if defined(__ARM_FEATURE_CRYPTO)
+    // Adapted from: https://mary.rs/lab/crc32/
+    // Barrent reduction
+    uint64x2_t orig =
+        vcombine_u64(vcreate_u64((uint64_t) (crc) << 24), vcreate_u64(0x0));
+    uint64x2_t tmp = orig;
+
+    // Polynomial P(x) of CRC32C
+    uint64_t p = 0x105EC76F1;
+    // Barrett Reduction constant (u64`) = u` = (x**64 / P(x))` = 0x11f91caf6 in
```
[the remark this answers is not in the document]
Comment (Cuda-Chen) [reply sse2neon.h]: Done.

File: sse2neon.h [hunk 6]
```diff
@@ -8500,12 +8500,47 @@ FORCE_INLINE uint32_t _mm_crc32_u8(uint32_t crc, uint8_t v)
...
+    tmp =
+        vandq_u64(tmp, vcombine_u64(vcreate_u64(0xFFFFFFFF), vcreate_u64(0x0)));
+    // Multiply by P(x) (shifted left by 1 for alignment reasons)
+    tmp = _sse2neon_vmull_p64(vget_low_u64(tmp), vcreate_u64(p));
+    // Subtract original from result
+    tmp = veorq_u64(tmp, orig);
+
+    // Extract the 'lower' (in bit-reflected sense) 32 bits
+    crc = vgetq_lane_u32(vreinterpretq_u32_u64(tmp), 1);
+#else
```
Comment (jserv) [inline sse2neon.h:8528]: Add comment "Fall back to the generic table lookup approach"

Comment (Cuda-Chen) [reply sse2neon.h]: Done.

**Comment by jserv** [comment]
> before converting this PR to a normal PR, I would like to ask whether it is necessary to implement runtime dispatching in this PR?

Compile-time options for diverse Arm extensions would be great.

**Comment by Cuda-Chen** [comment]
> > before converting this PR to a normal PR, I would like to ask whether it is necessary to implement runtime dispatching in this PR?
> 
> Compile-time options for diverse Arm extensions would be great.

For my experience, I will use `-mcpu` to determine the instruction set and the features (e.g., crypto and crc).

Anyway, I think I will make this PR ready for merge.

**Pull request merged**